### ISSUE TITLE
SPAM calculation not yet enabled

### ISSUE THREAD

**Issue by julimox** [Amber-MD/cpptraj#363]
Hello everybody,

I want to use SPAM and the calculation seems to run fine and also I get some output. However I doubt whether I can really trust it because during the calculations I get the message:

ACTION OUTPUT:
Warning: SPAM: SPAM calculation not yet enabled.

How can I enable SPAM?
Thank you very much!

**Comment by drroe** [comment]
I'm going to ping @swails on this since he's the author of the SPAM code within cpptraj and can correct anything I say that is wrong.

My understanding is that the SPAM code in cpptraj is functional, but is slightly different from the original implementation in that it uses a simpler non-bonded calculation incorporating a force-/energy-shifted electrostatic calculation instead of full PME. The procedure is to first run the `volmap` command with the `peakfile` keyword in order to generate the density peaks needed by `spam`. Then you can run the `spam` action and post-process those results: see details in a post from @swails here: http://archive.ambermd.org/201401/0290.html

I think in principle SPAM could be fully implemented in cpptraj now since it has access to PME energies via the sander API, but that's a decent amount of work and I don't have a timeline for when it might be done.

**Comment by julimox** [comment]
Thank you. I am using the python script he has uploaded in the same post and amber16. 
Let's see what he answers. Maybe there is a better way of using Spam now?

**Comment by hainm** [comment]
> and I don't know how to install this chemistry module

there is no module name `chemistry` in AT16. It's renamed to `parmed`.

**Comment by hainm** [comment]
So my suggestion is try to grep 'chemistry' and change it to 'parmed'. But let wait for Jason's feedback first.

**Comment by swails** [comment]
@drroe has the situation pretty well described.  The energy evaluation in cpptraj is a simple energy-shifted nonbonded calculation, whereas the original implementation utilized 3 PME calculations within NAMD to give the energy of each "bound" water.  In the test systems published in the original SPAM paper, the results between the current cpptraj energy calculations and those by NAMD were pretty small (for example, one water position's energy distribution changed from ~-1 to -30 kcal/mol to +2 to 32 kcal/mol).  The averages were basically the same, but the energy shifting function gave a little more spread.

@hainm's suggestion for replacing `chemistry` with `parmed` and installing the ParmEd package (either through AmberTools 16 or the Github repository) are the same suggestions I would make.

**Comment by drroe** [comment]
@swails so should the following warning be removed (or somehow modified)?
```
Warning: SPAM: SPAM calculation not yet enabled.
```

**Comment by swails** [comment]
Yea, I would say so.

**Comment by julimox** [comment]
Now it works =)
Thanks a lot!

**Comment by julimox** [comment]
I believe to have SPAM installed correctly now. Also I realise that the error occurs while I am not yet using SPAM.py but still within cpptraj. 

I run $AMBERTOOLS/cpptraj -i spam.in
With spam.in
parm <name>.prmtop
trajin <name>.mdcrd
spam peaks solv SOL reorder name <name>_spam bulk -30 info spam_info.out summary spam_sum.out out spam.out

Without "summary" - meaning "no SPAM energies will be calculated" I get no error message. 
But with the "summary" flag I still get 
ACTION OUTPUT:
Warning: SPAM: SPAM calculation not yet enabled.

**Comment by swails** [comment]
Oh yea...  The SPAM calculation that's referring to is the actual aggregation of the interaction energy data sets and calculating free energies, enthalpies, and entropies of those.

I hadn't gotten around to actually coding that calculation.  But the hard part of SPAM is the energy calculation.  If you can grok Python code, you can read the statistics module in SPAM.py to find out how to convert the energy data into SPAM results (it's not too bad).  Just leave out the "summary" keyword to generate the raw data.

**Comment by julimox** [comment]
So I should run $AMBERTOOLS/cpptraj spam peaks solv SOL reorder name name_spam bulk -30 info spam_info.out out spam.out

> Oh yea... The SPAM calculation that's referring to is the actual aggregation of the interaction energy data sets and calculating free energies, enthalpies, and entropies of those.
> 
> I hadn't gotten around to actually coding that calculation. But the hard part of SPAM is the energy calculation. If you can grok Python code, you can read the statistics module in SPAM.py to find out how to convert the energy data into SPAM results (it's not too bad). Just leave out the "summary" keyword to generate the raw data.

I am confused. I thought SPAM.py was converting the energy data!

**Comment by swails** [comment]
Originally, SPAM was implemented by adding a routine to cpptraj to identify water density peaks and reorder the trajectory file (and that was all it did).  SPAM.py then took this reordered trajectory and passed it through NAMD (using pairInteraction) to extract the energies for each water molecule in each peak location (for each frame), using the cpptraj-reordered trajectory.  Then, SPAM.py would read those energies and process them into G, H, and S.

The goal was to replace SPAM.py entirely within cpptraj so that the Python script wasn't necessary.  I got as far as implementing the energy calculation in cpptraj (so that NAMD was no longer necessary), but didn't get around to aggregating the energy sets into G, H, and S.

So now it's no longer necessary to get NAMD up and running -- but you still need to aggregate the SPAM energies from cpptraj into a measure of G, H, and S for each water.  My suggestion was to use the code in SPAM.py as an example of how that should be done.

**Comment by julimox** [comment]
I assume you refer to the function: spam_energies. 
Ok - I will try to do it!

**Comment by julimox** [comment]
Hello Jaison,
In your publication I have seen how you calculated the energies and during the last days I have been trying to get something like this in python but there is no hope. Is there any chance you will be working on SPAM any time soon?

**Comment by swails** [comment]
> In your publication I have seen how you calculated the energies and during the last days I have been trying to get something like this in python but there is no hope.

Have you tried using the Python code in the original SPAM.py package?  I posted it to GitHub so you can see how to code that here: https://github.com/swails/spam/blob/master/spam/spamstats.py#L14-L57

What I would suggest doing is to clone that repository and do the following:

```python
from spam.spamstats import calc_g_wat

# Read in the time series of your energies for each water site, then loop through each water site and
# run the following:

for site in sites: # here I assume site is a 1-D numpy array with an energy for each snapshot for 1 water
    gavg, gstd, havg, hstd, neg_entropy_t = calc_g_wat(site, 0, 10)
```

That will give you the average and standard deviations of the free energy, enthalpy, and -T*S (in kcal/mol).  The last two arguments are the sample size (0 means sample size is total number of samples and 10 means you want 10 subsamples to do bootstrap analysis on).

All you have to do before this is to run cpptraj to get the energy vectors for each site and then read them into numpy arrays.

> Is there any chance you will be working on SPAM any time soon?

Unlikely.  I no longer work in this field, and SPAM is too low on my priority list to spend any of my spare time on.  Who knows what the future holds, but this certainly won't happen in the foreseeable future.

**Comment by drroe** [comment]
Maybe I'll take a crack at this soon. @swails I know your time is limited, but do you have some sort of test case for `calc_g_wat` you can point me to so I can make sure I don't screw up the results?

**Comment by swails** [comment]
Not that I still have access to...  It should be easy enough to invent a time series and run it through `calc_g_wat` to make sure that cpptraj gives the same result for that calculation.

The constants that are hard-coded into that Python code are, I believe, input parameters to the spam action.

**Comment by julimox** [comment]
Do you mean the output of spam action? Which contains the energies for each peak per frame? I could give you this file if you want!

**Comment by drroe** [comment]
@swails I've been working on getting the final energy calcs into cpptraj. I'm seeing some differences that I'm having trouble resolving. Below are PDFs calculated from spamstats.py and cpptraj for a single water site.
![dbgkde 0](https://cloud.githubusercontent.com/assets/10144183/20761951/ae0fc2a2-b6f2-11e6-912d-92cf6a680d0b.jpg)
As you can see they are very similar. However, they end up giving different free energy values (on the order of 0.88 kcal/mol on average). After some trial and error, I've narrowed it down to this in spamstats.py:
```python
      free_energy[ii] = -1.373 * log10(binwidth *
                                      np.sum(kdevals * np.exp(-pts / 0.596)))
```
If I remove the `binwidth` factor from the logarithm, I get much better agreement (on the order of 0.08 kcal/mol on average). Do you remember why that `binwidth` factor is in there? I can't seem to puzzle it out - it seems like it's unnecessary to me.

**Comment by swails** [comment]
If I recall correctly, I took that from Guanglei.  This is a bit of a blast from the past :).  It took me a sec to realize where the 50 comes from in assigning "pts" (it does 50 points on either end of the min and max to capture the tails of the distributions).

But looking at this now, it looks to me like `binwidth * np.sum(kdevals * np.exp(-pts / 0.596))` is the partition function evaluated via an approximate Riemann sum, where the binwidth is the distance between the rectangles.

Admittedly it's been awhile since I've looked at this closely and I'm not sure about it.

**Comment by drroe** [comment]
OK - I think I see - it's because the summation in equation 2 of the SPAM paper is really a stand-in for integration, so in effectively is a left Riemann sum. In that case I wonder if it might be better to employ the trapezoidal rule.

**Comment by swails** [comment]
My approach was just make the rectangles skinny so I didn't have to look for, or implement, something harder than a simple left- or right-Riemann sum (e.g., left *and* right sums) :).

Or we could go whole hog and do Simpson's rule :)

**Comment by swails** [comment]
In seriousness though, if you notice an actual difference in accuracy, then by all means use a more accurate summation method.  I tend to be a bit sloppier when I numerically integrate a KDE-smoothed curve, since I can make the rectangles in the sum arbitrarily thin.

**Comment by drroe** [comment]
It seems OK the way it is. If needed I can re-assess in the future.

**Comment by drroe** [comment]
@julimox Sorry for the long delay on this, but everything should be working now. The SPAM calculation will actually run much faster if you have access to multiple cores/nodes as it works with OpenMP/MPI. If you get a chance to try it I would appreciate any feedback.

**Comment by julimox** [comment]
Great!! =)

How should I update my code? I assume the new SPAM is not updated to Amber
yet?

**Comment by swails** [comment]
You'll need to clone this repository and build cpptraj by itself.  It should be pretty straightforward to do so (just make sure you use *this* cpptraj instead of the one with AmberTools)

**Comment by swails** [comment]
One comment I'll make is in regards to how the energy calculation is done.  In the original implementation (a clunky workflow that only Guanglei could execute), the energies were computed via PME through NAMD's PairInteraction functionality computing the interaction between a single water molecule and every other atom in the system.

What this did was essentially 3 separate PME calculations:

1. PME of the water molecule with itself
2. PME of the rest of the system excluding the water molecule with itself
3. PME of everything

The energy is then computed as 3 - 2 - 1.  Since PME was not available within cpptraj when I added the precursor to this action several years ago, I used a simple energy shifting function instead.  What I found was that the energy distributions you get with the shifting function have the same mean, but a little broader distribution (maybe 5-10% broader) than what you get using PME.  I never did an in-depth study about the net effect of this broader distribution on SPAM results, but I would expect it to make the entropy contributions a little larger.
